imageregistry/v1/types.go

```diff
@@ -306,6 +306,36 @@ type ImageRegistryConfigStorageIBMCOS struct {
 	ServiceInstanceCRN string `json:"serviceInstanceCRN,omitempty"`
 }
 
+// ImageRegistryConfigStorageOSS holds Alibaba Cloud OSS configuration.
+type ImageRegistryConfigStorageOSS struct {
+	// bucket is the bucket name in which you want to store the registry's
+	// data.
+	// Optional, will be generated if not provided.
+	// +optional
+	Bucket string `json:"bucket,omitempty"`
+	// region is the GCS location in which your bucket exists.
+	// Optional, will be set based on the installed GCS Region.
+	// +optional
+	Region string `json:"region,omitempty"`
+	// regionEndpoint is the endpoint for S3 compatible storage services.
+	// Optional, defaults based on the Region that is provided.
+	// +optional
+	RegionEndpoint string `json:"regionEndpoint,omitempty"`
+	// internal specifies whether the registry use the OSS VPC internal endpoint
+	// Optional, defaults to false. if RegionEndpoint is specified, this config will be ignored
+	// +optional
+	Internal bool `json:"internal,omitempty"`
+	// encrypt specifies whether the registry stores the image in encrypted
+	// format or not.
+	// Optional, defaults to false.
+	// +optional
+	Encrypt bool `json:"encrypt,omitempty"`
```

Comment:
You have two encryption related fields?  this and keyID?  You need to describe the interaction and likely produce an API that makes it impossible to specify an invalid combination.

Reply:
Why would we allow a non-encrypted option?

Reply:
Here the documation https://www.alibabacloud.com/help/doc-detail/117914.htm?spm=a2c63.p38356.b99.1075.5c3e56989tiYEz. 
Encrypt means enable server-side encryption.  The default server-side encryption method.
Valid values: KMS, AES256. So if you not set keyID ,AES256 is used for default server-side encryption method.

Reply:
keep same as. https://github.com/docker/docker.github.io/blob/master/registry/storage-drivers/oss.md. 
![image](https://user-images.githubusercontent.com/2653769/133541562-daf42e36-f40e-4c06-b7bb-bdc74aa076f3.png)

Reply:
The registry storage driver isn't an openshift API.  This API is an openshift API and should conform to our standards.  `Encryption` with values of "ServerSideEncryption" and "ClearText" is very clear.

Reply:
Is it consistent with the AWS definition? Any suggestions?
type ImageRegistryConfigStorageS3 struct {
       .... 
	Bucket string `json:"bucket,omitempty"`
	// region is the AWS region in which your bucket exists.
	// Optional, will be set based on the installed AWS Region.
	// +optional
	Region string `json:"region,omitempty"`
	// regionEndpoint is the endpoint for S3 compatible storage services.
	// Optional, defaults based on the Region that is provided.
	// +optional
	RegionEndpoint string `json:"regionEndpoint,omitempty"`
	// encrypt specifies whether the registry stores the image in encrypted
	// format or not.
	// Optional, defaults to false.
	// +optional
	Encrypt bool `json:"encrypt,omitempty"`
	// keyID is the KMS key ID to use for encryption.
	// Optional, Encrypt must be true, or this parameter is ignored.
	// +optional
	KeyID string `json:"keyID,omitempty"`

Reply:
We cannot fix previous mistakes (this was in an operator repo when it was created), but we can avoid creating new instances of it.

Reply:
> The registry storage driver isn't an openshift API. This API is an openshift API and should conform to our standards. Encryption with values of "ServerSideEncryption" and "ClearText" is very clear.

You means we need to change `Encrypt bool` to   `Encryption string` ? Sorry, I don't understand it very clearly.

Reply:
@menglingwei I believe David is asking for an Enum type. Generally you create your own type which is a string and then reference that new type in place of the boolean. I'll let David correct me if I am wrong.

Reply:
@kwoodson @deads2k Is it like below?

```
type Encryption string
Encrypt Encryption `json:"encrypt,omitempty"`
```

But I think Encrypt means is ServerSideEncryption enabled? Would it be better if I changed it to a pointer type? 
```
Encrypt *bool json:"encrypt,omitempty"
```
If Encrypt not set, we can set default value to true.

Reply:
@menglingwei I believe we would like to remove the boolean types in favor of enumerables strings like this example:
https://github.com/openshift/api/pull/1045/files#diff-62f72cae4377f69081ae0e696139679ea6c1939ab0459ccbb96e8d280d417ae4R31-R45

When the API is called for Alibaba, we can check if `var.encrypt == imageregistry.ServerSideEncryption` and pass the boolean to the API. This makes defaulting and being very specific about what is being passed.

Reply:
@kwoodson @deads2k 

According to the above description, the complete structure definition may be like below,is ok?

```
type EndpointAccessibility string
var (
   InternalEndpoint = EndpointAccessibility("Internal")
   PublicEndpoint = EndpointAccessibility("Public")
)

type Encryption string
var(
 ServerSideEncryption = Encryption("ServerSideEncryption")
)

type ImageRegistryConfigStorageOSS struct {
	// bucket is the bucket name in which you want to store the registry's data.
	// Optional, will be generated if not provided.
	// +optional
	// About Bucket naming, more details you can look at the [official documentation](https://www.alibabacloud.com/help/doc-detail/257087.htm)
	Bucket string `json:"bucket,omitempty"`
	// region is the Alibaba Cloud Region in which your bucket exists.
	// Optional, will be set based on the installed Alibaba Cloud Region.
	// +optional
	// For a list of regions, you can look at the [official documentation](https://www.alibabacloud.com/help/doc-detail/31837.html).
	Region string `json:"region,omitempty"`
	// EndpointAccessibility specifies whether the registry use the OSS VPC internal endpoint
	// Optional, defaults to Internal.
	// +optional
	EndpointAccessibility EndpointAccessibility `json:"endpointAccessibility,omitempty"`
	// encrypt specifies whether the registry stores the image in encrypted
	// format or not.
	// Optional, defaults to false.
	// +optional
	// More details, you can look cat the [official documentation](https://www.alibabacloud.com/help/doc-detail/117914.htm)
	Encrypt Encryption `json:"encrypt,omitempty"`
}
```

But i hava another questions, If user want to use KMS encrypt data, how to pass KMSID to api?

Reply:
User would set `Encrypt` to `imageregistry.AlibabaEncryptionEnabled` and then pass the `KeyId` to the `<kms keyid>`. Would that work?

In the registry where you set your `Config` (https://github.com/openshift/cluster-image-registry-operator/pull/724/files#diff-e0f51602f2f4156d47e688910139da09ff81fe2ac10efa43d7d453b06ee08ec8R52) object should check and validate these fields are being used.

Reply:
> User would set `Encrypt` to `imageregistry.AlibabaEncryptionEnabled` and then pass the `KeyId` to the `<kms keyid>`. Would that work?
> 
> In the registry where you set your `Config` (https://github.com/openshift/cluster-image-registry-operator/pull/724/files#diff-e0f51602f2f4156d47e688910139da09ff81fe2ac10efa43d7d453b06ee08ec8R52) object should check and validate these fields are being used.

So in the api defination, we may define KMSID ?

Reply:
@menglingwei If the `KMSID` will be used then please add it. I see that there is a `KeyID` field in the `S3` storage section. If you require a field for this attribute then please add it.

Reply:
@kwoodson @deads2k From the above discussion.  I think i need to change internal to an Enum filed like "EndpointAccessibility" , it may be Internal or Public, the default value is Internal. And the Encrty change to an Enum filed like Encryption,It is used to define the server-side encryption algorithm,  KMS or  AES256, and same time ,add KeyID , if the Encryption value is KMS, the KeyID must be set. So the final structure should be 

```

type EndpointAccessibility string
var (
   InternalEndpoint = EndpointAccessibility("Internal")
   PublicEndpoint = EndpointAccessibility("Public")
)

type Encryption string
var(
 AES256 = Encryption("AES256")
 KMS = Encryption("KMS")
)

type ImageRegistryConfigStorageOSS struct {
	// bucket is the bucket name in which you want to store the registry's data.
	// Optional, will be generated if not provided.
	// +optional
	// About Bucket naming, more details you can look at the [official documentation](https://www.alibabacloud.com/help/doc-detail/257087.htm)
	Bucket string `json:"bucket,omitempty"`
	// region is the Alibaba Cloud Region in which your bucket exists.
	// Optional, will be set based on the installed Alibaba Cloud Region.
	// +optional
	// For a list of regions, you can look at the [official documentation](https://www.alibabacloud.com/help/doc-detail/31837.html).
	Region string `json:"region,omitempty"`
	// EndpointAccessibility specifies whether the registry use the OSS VPC internal endpoint
	// Optional, defaults to Internal.
	// +optional
	EndpointAccessibility EndpointAccessibility `json:"endpointAccessibility,omitempty"`
	// encrypt specifies whether the registry stores the image in encrypted
	// format or not.
	// Optional, defaults to false.
	// +optional
	// More details, you can look cat the [official documentation](https://www.alibabacloud.com/help/doc-detail/117914.htm)
	Encrypt Encryption `json:"encrypt,omitempty"`
	// keyID is the KMS key ID to use for encryption.
	// +optional
	KeyID string `json:"keyID,omitempty"`
}
```

I wonder if my understanding is correct? If it's correct, I'll do it this way

Reply:
If user don't set Encrypt or out of range ,the default value is AES256. If user set Encrypt to KMS and KeyID is not empty, use KMS.  Follow the [api documents](https://www.alibabacloud.com/help/doc-detail/117914.htm?spm=a2c63.p38356.b99.1075.5c3e56989tiYEz)

type Encryption string
var(
 AES256 = Encryption("AES256")
 KMS = Encryption("KMS")
)

```
Encrypt Encryption `json:"encrypt,omitempty"`
KeyID string `json:"keyID,omitempty"`
```

Reply:
Got it.

Reply:
this thread  got resolved before a push, unresolving so we can see it.